File: frontend/src/translation/TranslationLayer.ts
/**
 * Frontend Translation Layer - Pure Bridge to Protobuf
 * Translates between internal chemical/mystical metaphors and pure protobuf schemas
 * Maintains ontological purity at communication boundaries
 */

// Pure protobuf interfaces (matching backend protobuf schemas)
export interface ProtobufSystemComponent {
  id: string;
  type: 'AGGREGATE' | 'TRANSFORM' | 'CONNECT' | 'GENERATE';
  metadata: Record<string, string>;
  timestamp: number;
  status: Record<string, any>;
}

export interface ProtobufComponentOperation {
  component_id: string;
  type: 'AGGREGATE' | 'TRANSFORM' | 'CONNECT' | 'GENERATE';
  operation_name: string;
  input_data: Uint8Array;
  parameters: Record<string, string>;
  timestamp: number;
}

export interface ProtobufOperationResult {
  success: boolean;
  error_message: string;
  output_data: Uint8Array;
  metrics: Record<string, string>;
  execution_time_ms: number;
  timestamp: number;
}

// Internal component interface (implementation components)
export interface ComponentInterface {
  id: string;
  type: string;
  metadata: Record<string, any>;
  created_at: Date;
  get_status(): Record<string, any>;
}

export interface TranslationMetrics {
  translations_performed: number;
  successful_translations: number;
  failed_translations: number;
  average_translation_time_ms: number;
  success_rate_percent: number;
  last_updated: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
  fidelity_score: number;
}

export class FrontendTranslationLayer {
  /**
   * Translation layer between chemical/mystical metaphor implementations and pure protobuf
   * 
   * This class maintains ontological purity by:
   * 1. Keeping chemical/mystical metaphors contained within frontend implementation
   * 2. Translating to pure protobuf schemas at boundaries
   * 3. Preserving all functional behavior while purifying communication
   */
  
  private translationMetrics: TranslationMetrics = {
    translations_performed: 0,
    successful_translations: 0,
    failed_translations: 0,
    average_translation_time_ms: 0.0,
    success_rate_percent: 0.0,
    last_updated: new Date().toISOString()
  };

  /**
   * Translate internal SacredComponent (chemical/mystical metaphors) to pure protobuf
   * 
   * Metaphor Translation Mappings:
   * - SacredComponent → SystemComponent
   * - "Sacred/Divine" terminology → "System" terminology
   * - Chemical/mystical metadata → Implementation metadata
   */
  translateToProtobuf(component: ComponentInterface): ProtobufSystemComponent {
    const startTime = performance.now();
    
    try {
      // Map component types from implementation to pure
      const typeMapping: Record<string, ProtobufSystemComponent['type']> = {
        // Pure components (new)
        'DataAggregator': 'AGGREGATE',
        'DataTransformer': 'TRANSFORM',
        'DataConnector': 'CONNECT',
        'DataGenerator': 'GENERATE',
        // Legacy components (for backward compatibility)
        'SacredAggregator': 'AGGREGATE',
        'SacredLambdaEngine': 'TRANSFORM',
        'SacredConnector': 'CONNECT',
        'SacredGenerator': 'GENERATE',
        'DivineLambdaEngine': 'TRANSFORM',
        'IonicAggregator': 'AGGREGATE',
        'QuantumConnector': 'CONNECT',
        'CosmicGenerator': 'GENERATE'
      };
      
      const componentType = component.type;
      const pureType = typeMapping[componentType] || 'AGGREGATE';
      
      // Translate metadata, removing metaphorical references
      const pureMetadata = this.purifyMetadata(component.metadata);
      
      // Get component status through existing interface
      const status = component.get_status();
      const pureStatus = this.purifyStatus(status);
      
      // Create pure protobuf representation
      const protobufComponent: ProtobufSystemComponent = {
        id: component.id,
        type: pureType,
        metadata: pureMetadata,
        timestamp: component.created_at.getTime(),
        status: pureStatus
      };
      
      // Update metrics
      const executionTime = performance.now() - startTime;
      this.updateTranslationMetrics(true, executionTime);
      
      return protobufComponent;
      
    } catch (error) {
      const executionTime = performance.now() - startTime;
      this.updateTranslationMetrics(false, executionTime);
      throw new Error(`Translation to protobuf failed: ${error}`);
    }
  }

  /**
   * Translate pure protobuf back to internal representation format
   * 
   * Note: This returns a plain object representation rather than recreating
   * the full SacredComponent to avoid circular dependencies
   */
  translateFromProtobuf(protobufComponent: ProtobufSystemComponent): Record<string, any> {
    const startTime = performance.now();
    
    try {
      // Map pure types back to implementation terminology (for internal use)
      const typeMapping: Record<string, string> = {
        'AGGREGATE': 'DataAggregator',
        'TRANSFORM': 'DataTransformer',
        'CONNECT': 'DataConnector',
        'GENERATE': 'DataGenerator'
      };
      
      const internalType = typeMapping[protobufComponent.type] || 'SacredComponent';
      
      // Restore internal metadata format
      const internalMetadata = this.restoreMetadata(protobufComponent.metadata);
      
      // Create internal representation
      const internalComponent = {
        id: protobufComponent.id,
        type: internalType,
        metadata: internalMetadata,
        created_at: new Date(protobufComponent.timestamp),
        status: protobufComponent.status
      };
      
      // Update metrics
      const executionTime = performance.now() - startTime;
      this.updateTranslationMetrics(true, executionTime);
      
      return internalComponent;
      
    } catch (error) {
      const executionTime = performance.now() - startTime;
      this.updateTranslationMetrics(false, executionTime);
      throw new Error(`Translation from protobuf failed: ${error}`);
    }
  }

  /**
   * Translate operation request to pure protobuf format
   */
  translateOperationToProtobuf(
    component: ComponentInterface,
    operationName: string,
    inputData: any,
    parameters: Record<string, any> = {}
  ): ProtobufComponentOperation {
    
    // Serialize input data
    let serializedInput: Uint8Array;
    if (typeof inputData === 'object') {
      serializedInput = new TextEncoder().encode(JSON.stringify(inputData));
    } else if (typeof inputData === 'string') {
      serializedInput = new TextEncoder().encode(inputData);
    } else if (inputData instanceof Uint8Array) {
      serializedInput = inputData;
    } else {
      serializedInput = new TextEncoder().encode(String(inputData));
    }
    
    // Convert parameters to string format
    const pureParameters: Record<string, string> = {};
    for (const [key, value] of Object.entries(parameters)) {
      pureParameters[key] = String(value);
    }
    
    // Map component type
    const typeMapping: Record<string, ProtobufComponentOperation['type']> = {
      // Pure components (new)
      'DataAggregator': 'AGGREGATE',
      'DataTransformer': 'TRANSFORM',
      'DataConnector': 'CONNECT',
      'DataGenerator': 'GENERATE',
      // Legacy components (for backward compatibility)
      'SacredAggregator': 'AGGREGATE',
      'SacredLambdaEngine': 'TRANSFORM',
      'SacredConnector': 'CONNECT',
      'SacredGenerator': 'GENERATE'
    };
    const pureType = typeMapping[component.type] || 'AGGREGATE';
    
    return {
      component_id: component.id,
      type: pureType,
      operation_name: operationName,
      input_data: serializedInput,
      parameters: pureParameters,
      timestamp: Date.now()
    };
  }

  /**
   * Translate operation result to pure protobuf format
   */
  translateResultToProtobuf(
    success: boolean,
    resultData: any = null,
    errorMessage: string = '',
    executionTimeMs: number = 0.0,
    metrics: Record<string, any> = {}
  ): ProtobufOperationResult {
    
    // Serialize result data
    let serializedOutput: Uint8Array;
    if (resultData !== null) {
      if (typeof resultData === 'object') {
        serializedOutput = new TextEncoder().encode(JSON.stringify(resultData));
      } else if (typeof resultData === 'string') {
        serializedOutput = new TextEncoder().encode(resultData);
      } else if (resultData instanceof Uint8Array) {
        serializedOutput = resultData;
      } else {
        serializedOutput = new TextEncoder().encode(String(resultData));
      }
    } else {
      serializedOutput = new Uint8Array(0);
    }
    
    // Convert metrics to string format
    const pureMetrics: Record<string, string> = {};
    for (const [key, value] of Object.entries(metrics)) {
      pureMetrics[key] = String(value);
    }
    
    return {
      success,
      error_message: errorMessage,
      output_data: serializedOutput,
      metrics: pureMetrics,
      execution_time_ms: executionTimeMs,
      timestamp: Date.now()
    };
  }

  /**
   * Remove metaphorical terms from metadata
   */
  private purifyMetadata(metadata: Record<string, any>): Record<string, string> {
    const pureMetadata: Record<string, string> = {};
    
    // Metaphorical to pure mappings
    const metaphorMappings: Record<string, string> = {
      'system_id': 'system_id',
      'component_type': 'component_type',
      'connection_strength': 'connection_strength',
      'structural_energy': 'structural_energy',
      'optimization_ratio': 'optimization_ratio',
      'processing_frequency': 'processing_frequency',
      'component_state': 'component_state',
      'system_alignment': 'system_alignment',
      'system_constant': 'system_constant',
      'data_connection': 'data_connection',
      'data_structure': 'data_structure',
      'structural_order': 'structural_order'
    };
    
    for (const [key, value] of Object.entries(metadata)) {
      // Map metaphorical terms to pure terms
      const pureKey = metaphorMappings[key] || key;
      
      // Ensure value is string for protobuf compatibility
      pureMetadata[pureKey] = String(value);
    }
    
    return pureMetadata;
  }

  /**
   * Remove metaphorical terms from status information
   */
  private purifyStatus(status: Record<string, any>): Record<string, any> {
    const pureStatus: Record<string, any> = {};
    
    for (const [key, value] of Object.entries(status)) {
      // Remove metaphorical prefixes/suffixes and normalize
      let pureKey = key.replace(/^system_/, 'system_');
      pureKey = pureKey.replace(/^component_/, 'component_');
      pureKey = pureKey.replace(/^connection_/, 'connection_');
      pureKey = pureKey.replace(/^data_/, 'data_');
      pureKey = pureKey.replace(/_metric$/, '_metric');
      pureKey = pureKey.replace(/_connection$/, '_connection');
      
      pureStatus[pureKey] = value;
    }
    
    return pureStatus;
  }

  /**
   * Restore internal metadata format from pure protobuf
   */
  private restoreMetadata(pureMetadata: Record<string, string>): Record<string, any> {
    const internalMetadata: Record<string, any> = {};
    
    // Pure to internal mappings (identity mapping for pure components)
    const pureMappings: Record<string, string> = {
      'system_id': 'system_id',
      'component_type': 'component_type',
      'connection_strength': 'connection_strength',
      'structural_energy': 'structural_energy',
      'optimization_ratio': 'optimization_ratio',
      'processing_frequency': 'processing_frequency',
      'component_state': 'component_state',
      'system_alignment': 'system_alignment',
      'system_constant': 'system_constant',
      'data_connection': 'data_connection',
      'data_structure': 'data_structure',
      'structural_order': 'structural_order'
    };
    
    for (const [key, value] of Object.entries(pureMetadata)) {
      // Map pure terms back to metaphorical terms for internal use
      const internalKey = pureMappings[key] || key;
      internalMetadata[internalKey] = value;
    }
    
    return internalMetadata;
  }

  /**
   * Update translation performance metrics
   */
  private updateTranslationMetrics(success: boolean, executionTimeMs: number): void {
    this.translationMetrics.translations_performed += 1;
    
    if (success) {
      this.translationMetrics.successful_translations += 1;
    } else {
      this.translationMetrics.failed_translations += 1;
    }
    
    // Update average execution time
    const totalTranslations = this.translationMetrics.translations_performed;
    const currentAvg = this.translationMetrics.average_translation_time_ms;
    const newAvg = ((currentAvg * (totalTranslations - 1)) + executionTimeMs) / totalTranslations;
    this.translationMetrics.average_translation_time_ms = newAvg;
    
    // Update success rate
    this.translationMetrics.success_rate_percent = 
      (this.translationMetrics.successful_translations / totalTranslations) * 100;
    
    this.translationMetrics.last_updated = new Date().toISOString();
  }

  /**
   * Get translation layer performance metrics
   */
  getTranslationMetrics(): TranslationMetrics {
    return { ...this.translationMetrics };
  }

  /**
   * Validate translation fidelity
   */
  validateTranslation(
    originalComponent: any,
    protobufComponent: ProtobufSystemComponent
  ): ValidationResult {
    const validationResult: ValidationResult = {
      valid: true,
      errors: [],
      warnings: [],
      fidelity_score: 1.0
    };
    
    // Check ID preservation
    if (originalComponent.id !== protobufComponent.id) {
      validationResult.errors.push('ID not preserved during translation');
      validationResult.valid = false;
    }
    
    // Check type mapping
    const typeMapping: Record<string, string> = {
      // Pure components (new)
      'DataAggregator': 'AGGREGATE',
      'DataTransformer': 'TRANSFORM',
      'DataConnector': 'CONNECT',
      'DataGenerator': 'GENERATE',
      // Legacy components (for backward compatibility)
      'SacredAggregator': 'AGGREGATE',
      'SacredLambdaEngine': 'TRANSFORM',
      'SacredConnector': 'CONNECT',
      'SacredGenerator': 'GENERATE'
    };
    const expectedType = typeMapping[originalComponent.type];
    if (protobufComponent.type !== expectedType) {
      validationResult.errors.push(
        `Type mapping incorrect: expected ${expectedType}, got ${protobufComponent.type}`
      );
      validationResult.valid = false;
    }
    
    // Check metadata preservation
    const originalMetadataCount = Object.keys(originalComponent.metadata).length;
    const protobufMetadataCount = Object.keys(protobufComponent.metadata).length;
    if (Math.abs(originalMetadataCount - protobufMetadataCount) > 2) {
      validationResult.warnings.push('Significant metadata count difference');
      validationResult.fidelity_score -= 0.1;
    }
    
    return validationResult;
  }
}

// Global translation layer instance
export const frontendTranslator = new FrontendTranslationLayer();

/**
 * Convenience function for translating component to protobuf
 */
export function translateComponentToProtobuf(component: any): ProtobufSystemComponent {
  return frontendTranslator.translateToProtobuf(component);
}

/**
 * Convenience function for translating protobuf to internal format
 */
export function translateProtobufToComponent(protobufComponent: ProtobufSystemComponent): Record<string, any> {
  return frontendTranslator.translateFromProtobuf(protobufComponent);
}

// Legacy aliases for backward compatibility
export const translateSacredComponentToProtobuf = translateComponentToProtobuf;
export const translateProtobufToSacredComponent = translateProtobufToComponent;

/**
 * Get frontend translation layer metrics
 */
export function getFrontendTranslationMetrics(): TranslationMetrics {
  return frontendTranslator.getTranslationMetrics();
}

File: hive/translation_layer.py
"""
Backend Translation Layer - Pure Bridge to Protobuf
Translates between internal biological metaphors and pure protobuf schemas
Maintains ontological purity at communication boundaries
"""

from typing import Dict, Any, Optional, List
from dataclasses import dataclass, asdict
from datetime import datetime
import json
import uuid

# Import our existing components (with biological metaphors)
from .primitives import HiveComponent, Aggregate, Transformation, Connector, GenesisEvent

@dataclass
class ProtobufSystemComponent:
    """Pure protobuf SystemComponent representation"""
    id: str
    type: str  # "AGGREGATE", "TRANSFORM", "CONNECT", "GENERATE"
    metadata: Dict[str, str]
    timestamp: int
    status: Dict[str, Any]

@dataclass
class ProtobufComponentOperation:
    """Pure protobuf ComponentOperation representation"""
    component_id: str
    type: str
    operation_name: str
    input_data: bytes
    parameters: Dict[str, str]
    timestamp: int

@dataclass
class ProtobufOperationResult:
    """Pure protobuf OperationResult representation"""
    success: bool
    error_message: str
    output_data: bytes
    metrics: Dict[str, str]
    execution_time_ms: float
    timestamp: int

class BackendTranslationLayer:
    """
    Translation layer between biological metaphor implementations and pure protobuf
    
    This class maintains ontological purity by:
    1. Keeping biological metaphors contained within backend implementation
    2. Translating to pure protobuf schemas at boundaries
    3. Preserving all functional behavior while purifying communication
    """
    
    def __init__(self):
        self.translation_metrics = {
            "translations_performed": 0,
            "successful_translations": 0,
            "failed_translations": 0,
            "average_translation_time_ms": 0.0
        }
    
    def translate_to_protobuf(self, hive_component: HiveComponent) -> ProtobufSystemComponent:
        """
        Translate internal HiveComponent (biological metaphors) to pure protobuf
        
        Metaphor Translation Mappings:
        - HiveComponent → SystemComponent
        - "Hive" terminology → "System" terminology
        - Biological metadata → Implementation metadata
        """
        start_time = datetime.now()
        
        try:
            # Map component types from biological to pure
            type_mapping = {
                "Aggregate": "AGGREGATE",
                "Transformation": "TRANSFORM", 
                "Connector": "CONNECT",
                "GenesisEvent": "GENERATE"
            }
            
            component_type = type(hive_component).__name__
            pure_type = type_mapping.get(component_type, "UNSPECIFIED")
            
            # Translate metadata, removing biological references
            pure_metadata = self._purify_metadata(hive_component.metadata)
            
            # Get component status through existing interface
            status = hive_component.get_status()
            pure_status = self._purify_status(status)
            
            # Create pure protobuf representation
            protobuf_component = ProtobufSystemComponent(
                id=hive_component.id,
                type=pure_type,
                metadata=pure_metadata,
                timestamp=int(hive_component.created_at.timestamp() * 1000),
                status=pure_status
            )
            
            # Update metrics
            execution_time = (datetime.now() - start_time).total_seconds() * 1000
            self._update_translation_metrics(True, execution_time)
            
            return protobuf_component
            
        except Exception as e:
            execution_time = (datetime.now() - start_time).total_seconds() * 1000
            self._update_translation_metrics(False, execution_time)
            raise Exception(f"Translation to protobuf failed: {str(e)}")
    
    def translate_from_protobuf(self, protobuf_component: ProtobufSystemComponent) -> Dict[str, Any]:
        """
        Translate pure protobuf back to internal representation format
        
        Note: This returns a dictionary representation rather than recreating
        the full HiveComponent to avoid circular dependencies
        """
        start_time = datetime.now()
        
        try:
            # Map pure types back to biological terminology (for internal use)
            type_mapping = {
                "AGGREGATE": "Aggregate",
                "TRANSFORM": "Transformation",
                "CONNECT": "Connector", 
                "GENERATE": "GenesisEvent"
            }
            
            internal_type = type_mapping.get(protobuf_component.type, "Unknown")
            
            # Restore internal metadata format
            internal_metadata = self._restore_metadata(protobuf_component.metadata)
            
            # Create internal representation
            internal_component = {
                "id": protobuf_component.id,
                "type": internal_type,
                "metadata": internal_metadata,
                "created_at": datetime.fromtimestamp(protobuf_component.timestamp / 1000),
                "status": protobuf_component.status
            }
            
            # Update metrics
            execution_time = (datetime.now() - start_time).total_seconds() * 1000
            self._update_translation_metrics(True, execution_time)
            
            return internal_component
            
        except Exception as e:
            execution_time = (datetime.now() - start_time).total_seconds() * 1000
            self._update_translation_metrics(False, execution_time)
            raise Exception(f"Translation from protobuf failed: {str(e)}")
    
    def translate_operation_to_protobuf(self, 
                                      component: HiveComponent,
                                      operation_name: str,
                                      input_data: Any,
                                      parameters: Dict[str, Any] = None) -> ProtobufComponentOperation:
        """Translate operation request to pure protobuf format"""
        
        # Serialize input data
        if isinstance(input_data, dict):
            serialized_input = json.dumps(input_data).encode('utf-8')
        elif isinstance(input_data, str):
            serialized_input = input_data.encode('utf-8')
        elif isinstance(input_data, bytes):
            serialized_input = input_data
        else:
            serialized_input = str(input_data).encode('utf-8')
        
        # Convert parameters to string format
        pure_parameters = {}
        if parameters:
            for key, value in parameters.items():
                pure_parameters[key] = str(value)
        
        # Map component type
        component_type = type(component).__name__
        type_mapping = {
            "Aggregate": "AGGREGATE",
            "Transformation": "TRANSFORM",
            "Connector": "CONNECT", 
            "GenesisEvent": "GENERATE"
        }
        pure_type = type_mapping.get(component_type, "UNSPECIFIED")
        
        return ProtobufComponentOperation(
            component_id=component.id,
            type=pure_type,
            operation_name=operation_name,
            input_data=serialized_input,
            parameters=pure_parameters,
            timestamp=int(datetime.now().timestamp() * 1000)
        )
    
    def translate_result_to_protobuf(self,
                                   success: bool,
                                   result_data: Any = None,
                                   error_message: str = "",
                                   execution_time_ms: float = 0.0,
                                   metrics: Dict[str, Any] = None) -> ProtobufOperationResult:
        """Translate operation result to pure protobuf format"""
        
        # Serialize result data
        if result_data is not None:
            if isinstance(result_data, dict):
                serialized_output = json.dumps(result_data).encode('utf-8')
            elif isinstance(result_data, str):
                serialized_output = result_data.encode('utf-8')
            elif isinstance(result_data, bytes):
                serialized_output = result_data
            else:
                serialized_output = str(result_data).encode('utf-8')
        else:
            serialized_output = b""
        
        # Convert metrics to string format
        pure_metrics = {}
        if metrics:
            for key, value in metrics.items():
                pure_metrics[key] = str(value)
        
        return ProtobufOperationResult(
            success=success,
            error_message=error_message,
            output_data=serialized_output,
            metrics=pure_metrics,
            execution_time_ms=execution_time_ms,
            timestamp=int(datetime.now().timestamp() * 1000)
        )
    
    def _purify_metadata(self, metadata: Dict[str, Any]) -> Dict[str, str]:
        """Remove biological metaphors from metadata"""
        pure_metadata = {}
        
        # Biological to pure mappings
        biological_mappings = {
            "hive_id": "system_id",
            "organism_type": "component_type", 
            "dna_sequence": "config_sequence",
            "enzyme_function": "processor_function",
            "cell_membrane": "component_boundary",
            "nervous_system": "communication_layer",
            "mitochondria": "power_unit",
            "nucleus": "control_center",
            "ribosome": "synthesis_engine",
            "endoplasmic_reticulum": "transport_network",
            "golgi_apparatus": "packaging_system",
            "lysosome": "cleanup_mechanism"
        }
        
        for key, value in metadata.items():
            # Map biological terms to pure terms
            pure_key = biological_mappings.get(key, key)
            
            # Purify the value as well
            pure_value = str(value)
            for bio_term, pure_term in biological_mappings.items():
                if bio_term in pure_value:
                    pure_value = pure_value.replace(bio_term, pure_term)
            
            # Additional value purification
            pure_value = pure_value.replace("complex_cell", "complex_component")
            pure_value = pure_value.replace("_cell", "_component")
            pure_value = pure_value.replace("biological", "system")
            
            # Ensure value is string for protobuf compatibility
            pure_metadata[pure_key] = pure_value
        
        return pure_metadata
    
    def _purify_status(self, status: Dict[str, Any]) -> Dict[str, Any]:
        """Remove biological metaphors from status information"""
        pure_status = {}
        
        for key, value in status.items():
            # Remove biological prefixes/suffixes
            pure_key = key.replace("hive_", "system_")
            pure_key = pure_key.replace("_organism", "_component")
            pure_key = pure_key.replace("_dna", "_config")
            
            pure_status[pure_key] = value
        
        return pure_status
    
    def _restore_metadata(self, pure_metadata: Dict[str, str]) -> Dict[str, Any]:
        """Restore internal metadata format from pure protobuf"""
        internal_metadata = {}
        
        # Pure to biological mappings (reverse of purify)
        pure_mappings = {
            "system_id": "hive_id",
            "component_type": "organism_type",
            "config_sequence": "dna_sequence", 
            "processor_function": "enzyme_function",
            "component_boundary": "cell_membrane",
            "communication_layer": "nervous_system"
        }
        
        for key, value in pure_metadata.items():
            # Map pure terms back to biological terms for internal use
            internal_key = pure_mappings.get(key, key)
            internal_metadata[internal_key] = value
        
        return internal_metadata
    
    def _update_translation_metrics(self, success: bool, execution_time_ms: float):
        """Update translation performance metrics"""
        self.translation_metrics["translations_performed"] += 1
        
        if success:
            self.translation_metrics["successful_translations"] += 1
        else:
            self.translation_metrics["failed_translations"] += 1
        
        # Update average execution time
        total_translations = self.translation_metrics["translations_performed"]
        current_avg = self.translation_metrics["average_translation_time_ms"]
        new_avg = ((current_avg * (total_translations - 1)) + execution_time_ms) / total_translations
        self.translation_metrics["average_translation_time_ms"] = new_avg
    
    def get_translation_metrics(self) -> Dict[str, Any]:
        """Get translation layer performance metrics"""
        success_rate = 0.0
        if self.translation_metrics["translations_performed"] > 0:
            success_rate = (self.translation_metrics["successful_translations"] / 
                          self.translation_metrics["translations_performed"]) * 100
        
        return {
            **self.translation_metrics,
            "success_rate_percent": success_rate,
            "last_updated": datetime.now().isoformat()
        }
    
    def validate_translation(self, 
                           original_component: HiveComponent,
                           protobuf_component: ProtobufSystemComponent) -> Dict[str, Any]:
        """Validate translation fidelity"""
        validation_result = {
            "valid": True,
            "errors": [],
            "warnings": [],
            "fidelity_score": 1.0
        }
        
        # Check ID preservation
        if original_component.id != protobuf_component.id:
            validation_result["errors"].append("ID not preserved during translation")
            validation_result["valid"] = False
        
        # Check type mapping
        type_mapping = {
            "Aggregate": "AGGREGATE",
            "Transformation": "TRANSFORM",
            "Connector": "CONNECT",
            "GenesisEvent": "GENERATE"
        }
        expected_type = type_mapping.get(type(original_component).__name__)
        if protobuf_component.type != expected_type:
            validation_result["errors"].append(f"Type mapping incorrect: expected {expected_type}, got {protobuf_component.type}")
            validation_result["valid"] = False
        
        # Check metadata preservation
        original_metadata_count = len(original_component.metadata)
        protobuf_metadata_count = len(protobuf_component.metadata)
        if abs(original_metadata_count - protobuf_metadata_count) > 2:  # Allow some mapping differences
            validation_result["warnings"].append("Significant metadata count difference")
            validation_result["fidelity_score"] -= 0.1
        
        return validation_result

# Global translation layer instance
backend_translator = BackendTranslationLayer()

def translate_hive_component_to_protobuf(component: HiveComponent) -> ProtobufSystemComponent:
    """Convenience function for translating HiveComponent to protobuf"""
    return backend_translator.translate_to_protobuf(component)

def translate_protobuf_to_hive_component(protobuf_component: ProtobufSystemComponent) -> Dict[str, Any]:
    """Convenience function for translating protobuf to internal format"""
    return backend_translator.translate_from_protobuf(protobuf_component)

def get_backend_translation_metrics() -> Dict[str, Any]:
    """Get backend translation layer metrics"""
    return backend_translator.get_translation_metrics()